src/GraphQLParser/Visitors/SDLWriter.cs

```diff
@@ -0,0 +1,765 @@
+                    await Visit(node.Interfaces[i], context);
+                    if (i < node.Interfaces.Count - 1)
+                        await context.Writer.WriteAsync(" & ");
+                }
+            }
+        }
+
+        //TODO: remove
+        private async ValueTask Visit<T>(List<T>? nodes, TContext context, string delimiter, string start)
+            where T : ASTNode
+        {
+            if (nodes != null)
+            {
+                if (!string.IsNullOrEmpty(start))
+                    await context.Writer.WriteAsync(start);
+
+                for (int i = 0; i < nodes.Count; ++i)
+                {
+                    await Visit(nodes[i], context);
+                    if (i < nodes.Count - 1)
+                        await context.Writer.WriteAsync(delimiter);
+                }
+            }
+        }
+
+        private async ValueTask VisitDirectives(IHasDirectivesNode node, TContext context)
+        {
+            if (node.Directives?.Count > 0)
+            {
+                await context.Writer.WriteAsync(' ');
+
+                for (int i = 0; i < node.Directives.Count; ++i)
+                {
+                    await Visit(node.Directives[i], context);
+                    if (i < node.Directives.Count - 1)
+                        await context.Writer.WriteAsync(' ');
+                }
+            }
+        }
+
+        private string GetOperationType(OperationType type) => type switch
+        {
+            OperationType.Mutation => "mutation",
+            OperationType.Subscription => "subscription",
+            _ => "query",
+        };
+
+        private async ValueTask WriteIndent(TContext context, int level)
+        {
+            for (int i = 0; i < level; ++i)
+                await context.Writer.WriteAsync("  ");
+        }
+
+        private int GetLevel(TContext context)
+        {
+            int level = 0;
+
+            if (context.Parent.Count > 0)
+            {
+                var currentNode = context.Parent.Pop();
+
+                foreach (var node in context.Parent)
+                {
+                    if (node is GraphQLSelectionSet ||
+                        node is GraphQLTypeDefinition ||
+                        node is GraphQLInputObjectTypeDefinition ||
+                        node is GraphQLSchemaDefinition)
+                        ++level;
+                }
+
+                if (currentNode is GraphQLDescription)
+                    --level;
+                else if (currentNode is GraphQLComment && context.Parent.Peek() is GraphQLTypeDefinition)
+                    --level;
+
+                context.Parent.Push(currentNode);
+            }
+
+            return level;
+        }
+
+        private ValueTask Write(TContext context, ROM value)
+        {
+            var task =
+#if NETSTANDARD2_0
+            context.Writer.WriteAsync(value.ToString());
+#elif NETSTANDARD2_1_OR_GREATER
+            context.Writer.WriteAsync(value);
```

Comment:
```suggestion
            context.Writer.WriteAsync(value, context.CancellationToken);
```

Reply:
Cancellation tokens are supported here

Reply:
Might want to add a `private ValueTask WriteLine(TContext context)` method so the writeline methods can utilize the cancellation token.  (Store `Environment.NewLine` as a ROM statically, etc)

Reply:
And maybe a `private ValueTask WriteLine(char char)` method.....

Reply:
😄 I read these comments after I fixed all that stuff. Check again `WriteContextExtensions`.
Regarding writing single char. I rejected a separate method and pass a string from one symbol. Not such a big trouble, but the code is slightly clearer and unified.